### ISSUE TITLE
term hashmap memory usage

### ISSUE THREAD

**Issue by PSeitz** [quickwit-oss/tantivy#958]
I am writing a insert only string hashmap [inohash](https://github.com/PSeitz/insert_only_string_hash_map), and did compare performance and memory allocations with the new dhat crate between fnv and tantivitys term hashmap.

The memory consumption by the term hashmap seems to be extraordinary high in my test case, not sure if this already considered a bug or if I am doing wrong.

https://github.com/PSeitz/insert_only_string_hash_map/tree/main/compare_allocations
    
 ```
Running target/debug/inohash
dhat: Total:     7,813,120 bytes in 150 blocks

Running target/debug/tant_termmap
dhat: Total:     28,672,320 bytes in 150 blocks

Running target/debug/fnv
dhat: Total:     11,219,930 bytes in 137,850 blocks
 ```

**Comment by fulmicoton** [comment]
I think your result is correct.
I dived into that code and tried to check where tantivy was inefficient there.
Here is a list.

Tantivy stores strings and value in an inline manner in some kind of memory arena.
The hash table itself contains {the address for the string in the memory arena, a 32 bit hash, and an unordered term id}.
(More on this later.)

The point of the 32 bit hash is to help with collision and resize.

1.  We resize every time the hash table reach an occupancy > 33%. This is extremely aggressive and against common practises. I think it was motivated by a benchmark... But it might be due to the naiveness of my hash table implementation
2. We keep a Vec<u32> with the position that are filled in the hash table. If I recall correctly this is perfectly silly. We don't care about iterating in order, so a bitset or using one bit of the hash would be an ok alternative.
3. The page of the memory arena are overkill for the test. The memory arena I use allocates memory in page of 1MB. The current test does not fill a single page, so there is a lot of waste here. Tantivy uses the same term hash map for all fields so the term cardinality is usually quite high. I don't think this is something we need to fix.
4. We keep an unordered term id, which is an autoincrement id for the term. First term newly inserted get value 1, etc.
The point is that as we populate terms, the user need to get some kind of id identifying the term uniquely. I currently store this in the table, but it could absolutely be move into the memory arena.

**Comment by fulmicoton** [comment]
Here is a commit that changes 1,2,3 and 4.
https://github.com/fulmicoton/insert_only_string_hash_map/commit/20e25e107291664a8bb90a5c8c475578a3c51a8b

**Comment by fulmicoton** [comment]
The main difference between  inohash and tantivy's term hash map is the fact that tantivy is inline the string keys and the values.

Because the strings are variable length, I deal with alignment issue not by padding but by doing an explicit `{load, store}_unaligned`. Maybe padding would be nicer (!?)

Anyway... the theoretical benefit is to avoid two cache miss.

Your current bench is a tad slower to surface the benefit. Here is what happens on a 1MB test file with a higher cardinality.

```
test tests::bench_fnv                            ... bench:     115,713 ns/iter (+/- 2,246)
test tests::bench_fnv_full                       ... bench:  11,080,164 ns/iter (+/- 2,231,844)
test tests::bench_hasmap                         ... bench:     104,621 ns/iter (+/- 3,070)
test tests::bench_hasmap_full                    ... bench:   7,288,173 ns/iter (+/- 1,020,968)
test tests::bench_hasmap_full_large_struct       ... bench:  11,283,343 ns/iter (+/- 1,932,267)
test tests::bench_tant_termmap                   ... bench:     119,546 ns/iter (+/- 7,891)
test tests::bench_tant_termmap_full              ... bench:   6,981,644 ns/iter (+/- 832,690)
test tests::bench_tant_termmap_full_large_struct ... bench:   7,921,733 ns/iter (+/- 908,791)

```

**Comment by PSeitz** [comment]
Nice improvements, memory consumption looks good now. Which test file are you using? Yes, the book test case does no reflect  upon the cardinality correctly for the majority of use cases.

>We keep an unordered term id, which is an autoincrement id for the term. First term newly inserted get value 1, etc.
>The point is that as we populate terms, the user need to get some kind of id identifying the term uniquely. I currently store this in the table, but it could absolutely be move into the memory arena.

Yes, it's the same for me, I assign an id and count the occurrences. Why is the id not part of the regular `value` in the hashmap?

> The page of the memory arena are overkill for the test. The memory arena I use allocates memory in page of 1MB. The current test does not fill a single page, so there is a lot of waste here. Tantivy uses the same term hash map for all fields so the term cardinality is usually quite high. I don't think this is something we need to fix.

I also thought about doing that, but the resulting ids per field would not be minimal in size, I should probably reconsider that.
Does that mean all terms are in one fst in tantivity?

>The main difference between inohash and tantivy's term hash map is the fact that tantivy is inline the string keys and the values.
>
>Because the strings are variable length, I deal with alignment issue not by padding but by doing an explicit {load, store}_unaligned. Maybe padding would be nicer (!?)

I noticed there are performance issues when accessing the strings in inohash, but did not look into it yet. I think padding should be easy to achieve an be probably a bit faster. Also accessing always blocks like 8, 16, 24 bytes can have a major impact. At least that's what I saw when optimizing lz4_flex.

**Comment by fulmicoton** [comment]
> Yes, it's the same for me, I assign an id and count the occurrences. Why is the id not part of the regular value in the hashmap?

No reason at all. This is a a bad design.

**Comment by fulmicoton** [comment]
> I also thought about doing that, but the resulting ids per field would not be minimal in size, I should probably reconsider that. Does that mean all terms are in one fst in tantivity?

No. It is one fst per field, but the term hash map is common to all fields.

**Comment by fulmicoton** [comment]
My test data was generated using this.
```
import random

def gen_word():
    d = random.randint(1, 7)
    return "word" + str(random.randint(0, 10**d))

with open("testfile", "w") as f:
    for i in range(0, 10_000_000):
        f.write(gen_word() + " ")
```

The resulting file is too long though, so i trimmed it at 5MB if I recall correctly.

**Comment by PSeitz** [comment]
> No reason at all. This is a a bad design.

But like this the termid could be varint compressed :)

> My test data was generated using this.
> ...

This looks more like id generation, which is an important use case, but I think we can expect that a big amount of tokens are duplicates.
For ids I always think there should be something smarter to additionally reduce the space, instead of putting it in a hashmap. If you can detect patterns in it, there could be big gains possible. In this case the pattern would be prefix + numbers only.

**Comment by fulmicoton** [comment]
> But like this the termid could be varint compressed :)

The "But" in your sentence suggests you might have misinterpreted my sentence. 
I meant the current design in tantivy a bad design.
I agree what you are offering is better.

**Comment by PSeitz** [comment]
The but was in favor of your design. 
Because it is just a byte array and not `Sized` types, it is possible to use compression like varint.
I did some more benchmarks and so far my conclusion is that the hash maps are all so fast, that memory consumption is probably more important than performance when using it as term storage.
